python/ray/tune/trainable.py

```diff
@@ -1021,3 +1022,12 @@ def build_config(self, config: Dict):
         new_config[STDOUT_FILE] = self._stdout_file
         new_config[STDERR_FILE] = self._stderr_file
         return new_config
+
+
+class ConvertibleToTrainable(abc.ABC):
+    """A class that can be converted to a ``tune.Trainable``."""
+
+    @abc.abstractmethod
+    def as_trainable(self) -> Type[Trainable]:
```

Comment:
I think we should just move `as_trainable` directly to the `Trainer` interface above. First of all `ConvertibleToTrainable` serves no other purpose and is a questionable naming convention. Also, we want users to implement this ABC, right? I want my interfaces to be as explicit as possible and not search through several mix-ins first.

Reply:
Maybe you intend to reuse this class, but right now I don't see any benefit to it.

Reply:
I think one of the main reasons here is that we allow a dependency of Ray Train on Ray Tune (as the execution engine), but not from Ray Tune to Ray Train. Tune has to be differentiate between different trainables - string references, actual Trainable classes, and Trainers, but because we can't import Ray Train trainers we use this common abstract `ConvertibleToTrainable` class instead.

Reply:
+1

Reply:
with a FunctionTrainer in place I'm ok with keeping this, but I still think it's an anti-pattern, for the record. 

@krfricke I don't see how your (otherwise very valid) argument applies in the concrete situation. Whether a `Trainer` implements a `ConvertibleToTuneTrainableFunctionConverterAdapter` or simply has an `as_trainable` method doing the same thing makes no difference for visibility and dependencies.